compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldDelegate.skiko.kt

```diff
@@ -0,0 +1,144 @@
+ * - When you make a single tap on a word, the caret moves to the end of this word.
+ * - If there’s a punctuation mark after the word, the caret is between the word and the punctuation mark.
+ * - If you tap on a whitespace, the caret is placed before the word. Same for many whitespaces in a row. (and punctuation marks)
+ * - If there’s a punctuation mark before the word, the caret is between the punctuation mark and the word.
+ * - When you make a single tap on the first half of the word, the caret is placed before this word.
+ * - If you tap on the left edge of the TextField, the caret is placed before the first word on this line. The same is for the right edge.
+ * - If you tap at the caret, that is placed in the middle of the word, it will jump to the end of the word.
+ * @param offset The current offset position.
+ * @param currentValue The current TextFieldValue.
+ * @param textLayoutResult The TextLayoutResultProxy representing the layout of the text.
+ * @param currentText The current text in the TextField.
+ * @return The desired cursor position after evaluating the given parameters.
+ */
+internal fun determineCursorDesiredOffset(
+    offset: Int,
+    currentValue: TextFieldValue,
+    textLayoutResult: TextLayoutResultProxy,
+    currentText: String
+): Int {
+    val caretOffsetPosition: Int
+    val previousCaretPosition = currentValue.selection.start
+    if (offset == previousCaretPosition) {
+        caretOffsetPosition = offset
+    } else if (textLayoutResult.isLeftEdgeTapped(offset)) {
+        val lineNumber = textLayoutResult.value.getLineForOffset(offset)
+        caretOffsetPosition = textLayoutResult.value.getLineStart(lineNumber)
+    } else if (textLayoutResult.isRightEdgeTapped(offset)) {
+        val lineNumber = textLayoutResult.value.getLineForOffset(offset)
+        caretOffsetPosition = textLayoutResult.value.getLineEnd(lineNumber)
+    } else if (isPunctuationOrSpaceTapped(offset, currentText)) {
+        val nextWordStartOffset =
+            findNextNonWhitespaceSymbolsSubsequenceStartOffset(offset, currentText)
+        caretOffsetPosition = nextWordStartOffset
+    } else if (textLayoutResult.isFirstHalfOfWordTapped(offset, currentText)) {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.start
+    } else {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.end
+    }
+    return caretOffsetPosition
+}
+
+private fun isPunctuationOrSpaceTapped(
+    caretOffset: Int,
+    currentText: String
+): Boolean {
+    /* From TextLayoutResultProxy.value.getWordBoundary(caretOffset) it is clear
+    * that for whitespace or punctuation mark method will return empty range.
+    * caretOffset can be between 0...currentText.length, so it should be somewhere between words.
+    * */
+    return currentText.isPunctuation(caretOffset) || currentText.isWhitespace(caretOffset)
```

Comment:
I'd avoid extra string extensions because it leads to multiple `codePointAt` calls with the same parameters